zerver/views/streams.py

```diff
@@ -75,12 +77,69 @@
 from zerver.lib.typed_endpoint import ApiParamConfig, PathOnly, typed_endpoint
 from zerver.lib.typed_endpoint_validators import check_color, check_int_in_validator
 from zerver.lib.user_groups import access_user_group_for_setting
-from zerver.lib.users import access_user_by_email, access_user_by_id
+from zerver.lib.users import access_user_by_email, access_user_by_id, access_user_common
 from zerver.lib.utils import assert_is_not_none
 from zerver.models import NamedUserGroup, Realm, Stream, UserProfile
 from zerver.models.users import get_system_bot
 
 
+def bulk_principals_to_user_profiles(
+    agent: UserProfile, principals: list[str] | list[int]
+) -> set[UserProfile]:
+    # principals is guaranteed to be non-empty and to have the same type of elements, int or str,
+    # so the following check is safe and enough.
+
+    # principlas are user emails.
+    if isinstance(principals[0], str):
+        # using .lower() for each email and
+        # annotate(lower_email=Lower("email")) in the queryset
+        # achieves the same result of doing email__iexact__in=emails but that is not supported.
+        user_emails = [email.strip().lower() for email in principals]  # type: ignore[union-attr] # email is str as explained above
+        users_profiles: QuerySet[Any]
+        users_profiles = (
+            UserProfile.objects.annotate(lower_email=Lower("email"))
+            .select_related(
+                "realm",
+                "realm__can_access_all_users_group",
+                "realm__can_access_all_users_group__named_user_group",
+                "realm__direct_message_initiator_group",
+                "realm__direct_message_initiator_group__named_user_group",
+                "realm__direct_message_permission_group",
+                "realm__direct_message_permission_group__named_user_group",
+                "bot_owner",
+            )
+            .filter(lower_email__in=user_emails, realm=agent.realm)
+        )
+        valid_users_emails = {user_profile.lower_email for user_profile in users_profiles}
+        all_users_exist = all(email in valid_users_emails for email in user_emails)
+
+    # principals are user ids
+    else:
+        users_profiles = UserProfile.objects.select_related(
+            "realm",
+            "realm__can_access_all_users_group",
+            "realm__can_access_all_users_group__named_user_group",
+            "bot_owner",
```

Comment:
Why does this use a different `select_related` list? 

Also, it would be a better use of our abstraction boundaries to put the core internal logic in a couple functions in `zerver/lib/users.py` with names like `bulk_access_user_by_email`, placed right after `access_user_by_email`.

Generally code outside that module should never directly do `UserProfile.objects....` calls.

Reply:
> Why does this use a different `select_related` list?

because it's the same `select_related`  by [get_user](https://github.com/zulip/zulip/blob/main/zerver/models/users.py#L1010), with `access_user_by_email()` being the caller

Reply:
> Why does this use a different `select_related` list?

The difference in the `select_related` is the same difference between `access_user_by_email` and `access_user_by_id`, I simply reused them

Reply:
Also I've just noticed that `access_user_by_email()` calls [get_user](https://github.com/zulip/zulip/blob/main/zerver/models/users.py#L1009), which caches each `UserProfile` for each call, that's only when we are dealing with `emails` and not `ids`, do we want that same cache ?

Reply:
> Also, it would be a better use of our abstraction boundaries to put the core internal logic in a couple functions in `zerver/lib/users.py` with names like `bulk_access_user_by_email`, placed right after `access_user_by_email`, and call the appropriate such functions in the if/else clauses here.

Do we want `bulk_access_user_by_email` and `bulk_access_use_by_id` to be self contained such that `access_user_common()` called inside each of them ?  or instead, and to save a couple of lines `access_user_common()` should be called inside their caller  `bulk_principals_to_user_profiles()` ?

Reply:
`access_user_common` should definitely be called in each. Otherwise, they are not correct self-contained functions, and anything else that calls them thinking a `bulk_access_*` function is safe will be violating security invariants.

Reply:
> and anything else that calls them thinking a `bulk_access_*` function is safe will be violating security invariants.

Yes, That really made sense to me thinking about it.